presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java

```diff
@@ -372,11 +401,33 @@ private List<TaskSource> getTaskSources(
                     TaskSource source = new TaskSource(remoteSource.getId(), ImmutableSet.of(split), ImmutableSet.of(Lifespan.taskWide()), true);
                     shuffleTaskSources.add(source);
                 }
+
+                Broadcast<?> broadcast = nativeTaskInputs.getBroadcastInputs().get(sourceFragmentId.toString());
+                if (broadcast != null) {
+                    List<BroadcastFileInfo> fileInfos =
+                            ((List<?>) broadcast.value()).stream()
+                                    .map(PrestoSparkSerializedPage.class::cast)
+                                    .map(prestoSparkSerializedPage -> PrestoSparkUtils.toSerializedPage(prestoSparkSerializedPage))
+                                    .map(serializedPage -> pagesSerde.deserialize(serializedPage))
+                                    // Extract filePath.
+                                    .flatMap(page -> IntStream.range(0, page.getPositionCount())
+                                            .mapToObj(position -> VarcharType.VARCHAR.getObjectValue(null, page.getBlock(0), position)))
+                                    .map(String.class::cast)
+                                    .map(filePath -> new BroadcastFileInfo(filePath))
```

Comment:
A split usually represents a part of a file or the whole file.  Instead of supplying all file metadata as part of `broadcastInfo` as parameter to a single split, should we be creating a `ScheduledSplit` per file,  ? 
We could even go as far as setting the path to be the actual `ws` path in the `Location` field of the Split like so ?

```Java
...
.map(filePath -> new ScheduledSplit(
    nextSplitId.getAndIncrement(), 
    remoteSource.getId(), 
    new Split(
        REMOTE_CONNECTOR_ID, 
        new RemoteTransactionHandle(), 
        new RemoteSplit(new Location(broadcastDirectory + filePath)))) // <<-- FILE PATH HERE ?
.collect(toImmutableList());
```

I can see that we are just following the `shuffleInfo` pattern of stuffing everything into a metadata object, but that is probably not very well designed.

Reply:
+1
Try to avoid branching in code as much as possible. `ScheduledSplit` should be able to hide away the split complexity or broadcast path or non-broadcast path.

Reply:
Good call. 
I have updated PR